### PR TITLE
Metrics for endpoints + make dummy metrics singleton instances

### PR DESCRIPTION
**Pull request by vietj**
Provides HttpClient endpoint monitoring providing per endpoint queing / connections / usage

### CODE REVIEW

**Comment by vietj** [comment]
there is an implementation for dropwizard after that will follow

File: src/main/java/io/vertx/core/http/impl/HttpClientImpl.java [hunk 1]
```diff
@@ -55,7 +55,6 @@
   private final ConnectionManager connectionManager;
   private final Closeable closeHook;
   private final SSLHelper sslHelper;
-  final HttpClientMetrics metrics;
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/HttpClientImpl.java]: Out of curiosity, was there a particular reason for moving the HttpClientMetrics field to the connection manager? Wondering why the move since the lifecycle was already managed here.

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java [hunk 2]
```diff
@@ -391,8 +393,8 @@ protected synchronized void handleClosed() {
       for (HttpClientRequestImpl req: requests) {
         metrics.requestReset(req.metric());
       }
-      if (requestForResponse != null) {
-        metrics.requestReset(requestForResponse.metric());
+      if (currentResponse != null) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: This is not strictly related to the goal of the PR, is it? It looks like a bug fix.

File: src/main/java/io/vertx/core/http/impl/ConnectionManager.java [hunk 3]
```diff
@@ -281,8 +294,14 @@ private void createNewConnection(Waiter waiter) {
...
     Waiter getNextWaiter() {
       Waiter waiter = waiters.poll();
+      if (waiter != null && ConnectionManager.this.metrics.isEnabled()) {
+        ConnectionManager.this.metrics.dequeueRequest(metric, waiter.metric);
+      }
       while (waiter != null && waiter.isCancelled()) {
         waiter = waiters.poll();
+        if (waiter != null && ConnectionManager.this.metrics.isEnabled()) {
+          ConnectionManager.this.metrics.dequeueRequest(metric, waiter.metric);
+        }
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/ConnectionManager.java]:
```java
Waiter waiter;
do {
  waiter = waiters.poll();
  if (waiter == null) break;
  if (ConnectionManager.this.metrics.isEnabled()) {
    ConnectionManager.this.metrics.dequeueRequest(metric, waiter.metric);
  }  
} while (waiter.isCancelled());
```

File: src/main/java/io/vertx/core/http/impl/ConnectionManager.java [hunk 4]
```diff
@@ -281,8 +294,14 @@ private void createNewConnection(Waiter waiter) {
      */
     Waiter getNextWaiter() {
       Waiter waiter = waiters.poll();
+      if (waiter != null && ConnectionManager.this.metrics.isEnabled()) {
+        ConnectionManager.this.metrics.dequeueRequest(metric, waiter.metric);
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/ConnectionManager.java]: Thinking out loud.

The SPI implementer has no way to determine if `dequeue` is called for a canceled metric? Maybe we don't care because we want to measure waiting time in the queue for both waiters who passed and who timed out.

File: src/main/java/io/vertx/core/metrics/impl/DummyVertxMetrics.java [hunk 5]
```diff
@@ -109,7 +111,9 @@ public boolean isMetricsEnabled() {
     return false;
   }
 
-  protected class DummyEventBusMetrics implements EventBusMetrics<Void> {
+  public static class DummyEventBusMetrics implements EventBusMetrics<Void> {
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/metrics/impl/DummyVertxMetrics.java]: Nice move on making all those subclasses public here. Dummy impl is useful if you want to partially implement the SPI without having to recompile your code when a new method is added.

File: src/test/java/io/vertx/test/core/MetricsTest.java [hunk 6]
```diff
@@ -546,6 +546,118 @@ public void testHttpClientName() throws Exception {
...
+    }
+    assertEquals(Collections.singleton("localhost:8080"), metrics.endpoints());
+    assertEquals(8, (int)metrics.queueSize("localhost:8080"));
+    assertEquals(5, (int)metrics.connectionCount("localhost:8080"));
+    ArrayList<Runnable> copy = new ArrayList<>(requests);
+    requests.clear();
+    copy.forEach(Runnable::run);
+    awaitLatch(responsesLatch);
+    waitUntil(() -> requests.size() == 5);
+    copy = new ArrayList<>(requests);
```
Comment (tsegismont) [inline src/test/java/io/vertx/test/core/MetricsTest.java]: The test looks good to me. I'm just curious how we know that a maximum of 5 requests can be handled at a time. Did we set that in config somewhere? Maybe I'm missing the obvious...

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java (hunk 2 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: bug fix yes - was calling endRequest in metrics twice when connection was closed after response sent.

Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: there is a test for this in this PR

File: src/main/java/io/vertx/core/http/impl/ConnectionManager.java (hunk 4 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ConnectionManager.java]: yes we don't care we dont keep track of queue/deque like we do for pool

File: src/main/java/io/vertx/core/http/impl/HttpClientImpl.java (hunk 1 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/HttpClientImpl.java]: I prefer to keep it our of the HttpClientImpl that does not care much about

File: src/main/java/io/vertx/core/metrics/impl/DummyVertxMetrics.java (hunk 5 above)
Comment (vietj) [inline src/main/java/io/vertx/core/metrics/impl/DummyVertxMetrics.java]: yes that's the point : hack something quickly

File: src/test/java/io/vertx/test/core/MetricsTest.java (hunk 6 above)
Comment (vietj) [inline src/test/java/io/vertx/test/core/MetricsTest.java]: that's the default client config : 5 max

File: src/main/java/io/vertx/core/http/impl/ConnectionManager.java (hunk 4 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ConnectionManager.java]: we could but it means more complexity, I don't think it's worth to have it

File: src/test/java/io/vertx/test/core/MetricsTest.java (hunk 6 above)
Comment (tsegismont) [inline src/test/java/io/vertx/test/core/MetricsTest.java]: OK, makes sense then.
